iceoryx_utils/include/iceoryx_utils/internal/cxx/forward_list.inl

```diff
@@ -0,0 +1,581 @@
+{
+    if (this->m_cList != rhs_citer.m_cList)
+    {
+        errorMessage(__PRETTY_FUNCTION__, " iterators of different list can't be compared");
+        std::terminate();
+    }
+    // index comparison, not user_class operator==
+    return (m_iterListNodeIdx == rhs_citer.m_iterListNodeIdx);
+}
+
+
+template <typename T, int64_t CAPACITY>
+const T& forward_list<T, CAPACITY>::const_iterator::operator*() const noexcept
+{
+    if (!m_cList->isValidElementIndex(m_iterListNodeIdx))
+    {
+        // terminate e.g. when trying to read e.g. from end()
+        errorMessage(__PRETTY_FUNCTION__, " malformed iterator");
+        std::terminate();
+    }
+    CListNodePointer nodePtr = m_cList->getNodePtrFromIdx(m_iterListNodeIdx);
+    return nodePtr->data;
+} // operator*
+
+
+template <typename T, int64_t CAPACITY>
+const T* forward_list<T, CAPACITY>::const_iterator::operator->() const noexcept
+{
+    return &operator*();
+}
+
+
+// const_iterator
+/*************************/
+
+/*************************/
+// init
+template <typename T, int64_t CAPACITY>
+void forward_list<T, CAPACITY>::init() noexcept
+{
+    // initial link empty-list elements
+    getNodePtrFromIdx(BEFORE_BEGIN_USED_INDEX)->nextIdx = INVALID_INDEX;
+
+    for (size_type i = BEFORE_BEGIN_FREE_INDEX; (i + 1u) < INTERNAL_CAPACITY; ++i)
+    {
+        getNodePtrFromIdx(i)->nextIdx = i + 1;
+    }
+    getNodePtrFromIdx(INTERNAL_CAPACITY - 1)->nextIdx = INVALID_INDEX;
+} // init
+
+
+/*************************/
+// list-node access
+template <typename T, int64_t CAPACITY>
+inline typename forward_list<T, CAPACITY>::ListNode* forward_list<T, CAPACITY>::getNodePtrFromIdx(size_type idx) const
+    noexcept
+{
+    return &(getNodePtr()[idx]);
+}
+
+template <typename T, int64_t CAPACITY>
+inline typename forward_list<T, CAPACITY>::ListNode* forward_list<T, CAPACITY>::getHeadOfFreeElementList() const
+    noexcept
+{
+    return getNodePtrFromIdx(BEFORE_BEGIN_FREE_INDEX);
+}
+
+template <typename T, int64_t CAPACITY>
+inline typename forward_list<T, CAPACITY>::ListNode* forward_list<T, CAPACITY>::getNodePtr() const noexcept
+{
+    return reinterpret_cast<ListNode* const>(const_cast<uint8_t(*)[CAPACITY + 2][sizeof(ListNode)]>(&m_data));
+}
+
+
+template <typename T, int64_t CAPACITY>
+inline bool forward_list<T, CAPACITY>::isValidIteratorIndex(size_type index) const noexcept
+{
+    return index < INVALID_INDEX;
+}
+
+template <typename T, int64_t CAPACITY>
+inline bool forward_list<T, CAPACITY>::isValidElementIndex(size_type index) const noexcept
+{
+    return (isValidIteratorIndex(index) && index != BEFORE_BEGIN_USED_INDEX && index != BEFORE_BEGIN_FREE_INDEX);
+}
+
+template <typename T, int64_t CAPACITY>
+inline void forward_list<T, CAPACITY>::errorMessage(const char* f_source, const char* f_msg) noexcept
+{
+    std::cerr << f_source << " ::: " << f_msg << std::endl;
+}
+
+namespace
+{
+template <typename U, typename V>
+bool operator!=(U lhs, V rhs) noexcept
```

Comment:
Can be done more specialized for just the desired iterators```template <typename T, int64_t CAPACITY> bool operator!=(const forward_list<T, CAPACITY>::const_iterator &lhs, const  forward_list<T, CAPACITY>::const_iterator &rhs);```, if I do not miss something...